iree/hal/metal/metal_kernel_library.mm

```diff
@@ -0,0 +1,135 @@
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/hal/metal/metal_kernel_library.h"
+
+#include "iree/base/memory.h"
+#include "iree/base/status.h"
+#include "iree/base/tracing.h"
+
+namespace iree {
+namespace hal {
+namespace metal {
+
+// static
+StatusOr<ref_ptr<MetalKernelLibrary>> MetalKernelLibrary::Create(
+    id<MTLDevice> device, ExecutableCachingModeBitfield mode,
+    const MetalExecutableDef& metal_executable_def) {
+  IREE_TRACE_SCOPE0("MetalKernelLibrary::Create");
+  if (!metal_executable_def.entry_points() || metal_executable_def.entry_points()->size() == 0) {
+    return InvalidArgumentErrorBuilder(IREE_LOC) << "No entry points defined";
+  }
+  if (!metal_executable_def.shader_sources() ||
+      metal_executable_def.shader_sources()->size() == 0) {
+    return InvalidArgumentErrorBuilder(IREE_LOC) << "No MSL source string present";
+  }
+
+  const auto& entry_points = *metal_executable_def.entry_points();
+  const auto& msl_sources = *metal_executable_def.shader_sources();
+
+  if (entry_points.size() != msl_sources.size()) {
+    return InvalidArgumentErrorBuilder(IREE_LOC)
+           << "Number of entry points and source strings mismatch";
+  }
+
+  // Compile each MSL source string into a MTLLibrary and get the MTLFunction for the entry point to
+  // build the pipeline state object.
+
+  absl::InlinedVector<id<MTLLibrary>, 1> libraries;
+  absl::InlinedVector<id<MTLFunction>, 1> functions;
+  absl::InlinedVector<id<MTLComputePipelineState>, 1> states;
+
+  MTLCompileOptions* msl_compile_options = [MTLCompileOptions new];
+  msl_compile_options.languageVersion = MTLLanguageVersion2_0;
+
+  auto cleanup = MakeCleanup([&]() {
+    for (id<MTLComputePipelineState> state : states) [state release];
+    for (id<MTLFunction> function : functions) [function release];
+    for (id<MTLLibrary> library : libraries) [library release];
+    [msl_compile_options release];
+  });
+
+  for (int i = 0; i < msl_sources.size(); ++i) {
+    @autoreleasepool {
+      NSError* error = nil;
+
+      NSString* shader_source = [NSString stringWithCString:msl_sources[i]->c_str()
+                                                   encoding:[NSString defaultCStringEncoding]];
+      id<MTLLibrary> library = [device newLibraryWithSource:shader_source
+                                                    options:msl_compile_options
+                                                      error:&error];
+      if (!library) {
+        NSLog(@"Failed to create MTLLibrary: %@", error);
+        NSLog(@"Original MSL source: %@", shader_source);
```

Comment:
Aren't these source strings potentially very, very large? Not sure we should be logging them in their entirety... maybe move under a debug `#ifdef`?

(multiple times in this file)

Reply:
I was following the "being as verbose as possible on errors" philosophy here. Typically when something goes wrong, we don't really care about performance anymore. Giving more content for fixing the issue is more important. But it's fine for me to wrap it with `#ifndef NDEBUG`.